dnf/package.py

```diff
@@ -153,6 +160,22 @@ def debugsource_name(self):
         src_name = self.source_name if self.source_name is not None else self.name
         return src_name + self.DEBUGSOURCE_SUFFIX
 
+    def get_header(self):
+        """
+        Returns the rpm header of the package if it is installed. If not
+        installed, returns None. The header is not cached and retrieved from
+        rpmdb on every call. In case of a failure (e.g. when the rpmdb changes
+        between loading the data and calling this method), raises an instance
+        of PackageNotFoundError.
+        """
+        if self.repoid != "@System":
+            return None
+
+        try:
+            return next(rpm.TransactionSet().dbMatch(0, self.rpmdbid))
```

Comment:
I checked rpm documentation and it is not descriptive at all. Please could you provide an information how you find out?

Reply:
I see that the inspiration was taken from yum `def _get_hdr(self):`.

Reply:
@pmatilai I would like to ask you if this is the best way how to get rpm header from RPMDB?

Reply:
Eww. You don't open a database connection, retrieve one item and close it for every item you retrieve.
Yum does *not* set a good example wrt rpm API usage.

Reply:
@j-mracek yes, that's what I did, I couldn't really make out anything from the documentation, this is the only thing that worked for me.

@pmatilai I think Jaroslav's question was more directed towards the `dbMatch(0, self.rpmdbid)` part, in particular, what does the `0` even mean? :slightly_smiling_face: Regarding creating the `TransactionSet`, I don't have a clue what overhead that entails. Do you have any suggestion on how to do it better? There's the `_ts` attribute of `dnf.base.Base`, but that's used for actually performing package operations in rpm, I'm not sure I should use that? Or create another `TransactionSet`?

Reply:
Oh, didn't even notice that dbMatch(0, ...). The first argument is the rpmdb index you're doing lookups on, one of the `RPMDBI_*` values or corresponding name. `rpm.RPMDBI_PACKAGES` is `0`, ie the one you're using, which contains the actual headers. Others are indexes to them.

And then there's the ts-mess dnf inherited from yum, kinda forgotten about that. I'm not sure how much of that remains, but the notion of "readonly ts" that yum used is total bollocks, it doesn't reflect any reality within rpm, especially in the way yum used it. Think of the transaction set as a per-installroot global handle through which the database and other properties are accessed, there should only one should exist per a given installroot.

Rpm will open the database implicitly on queries, but for better control you can also do this manually (ts.openDB(), ts.closeDB()). The reason it matters that in current incarnations, having the rpmdb open changes signal delivery somewhat so it's something people often try to avoid during eg downloads. OTOH if one actually *uses* the rpm signal system rather than try to avoid it in vain (which is what yum tried and didn't really succeed), it shouldn't matter much. In any case, opening and closing the db is not something you want to do on per package level, it should be done higher up.

Reply:
Oh and BTW, besides the db open/close itself being somewhat costly, doing that also kills rpm's own caches which is significant especially if header integrity checks are left on (yum always disabled them entirely which then lead to various other issues)

Reply:
@pmatilai thanks for the explanation. I've changed the code to use `base._ts`, which is the transaction used to do package operations. Reading through how it's used, seems it should work fine.